monitoring/asset_proof_collector.go

```diff
@@ -0,0 +1,100 @@
+package monitoring
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	numAssetProofsMetric = "asset_proofs"
+)
+
+// assetProofCollector is a Prometheus collector that exports proof related data
+// of all taproot assets.
+type assetProofCollector struct {
+	collectMx sync.Mutex
+
+	cfg      *PrometheusConfig
+	registry *prometheus.Registry
+
+	proofsVec *prometheus.GaugeVec
+}
+
+func newAssetProofCollector(cfg *PrometheusConfig,
+	registry *prometheus.Registry) (*assetProofCollector, error) {
+
+	if cfg == nil {
+		return nil, errors.New("asset proof collector prometheus cfg " +
+			"is nil")
+	}
+
+	if cfg.AssetStore == nil {
+		return nil, errors.New("asset proof collector asset store is " +
+			"nil")
+	}
+
+	return &assetProofCollector{
+		cfg:      cfg,
+		registry: registry,
+		proofsVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: numAssetProofsMetric,
+				Help: "Proofs of all assets",
+			},
+			[]string{"asset_name"},
+		),
+	}, nil
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once the
+// last descriptor has been sent.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetProofCollector) Describe(ch chan<- *prometheus.Desc) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	a.proofsVec.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetProofCollector) Collect(ch chan<- prometheus.Metric) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	ctxdb, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+
+	// Fetch all proofs.
+	proofs, err := a.cfg.AssetStore.FetchAssetProofs(ctxdb)
+	if err != nil {
+		log.Errorf("unable to fetch asset proofs: %v", err)
+		return
+	}
+
+	a.proofsVec.Reset()
+
+	proofMap := make(map[string]prometheus.Gauge)
+
+	// We'll identify each gauge by the proof's script key.
+	for scriptKey, _ := range proofs {
+		key := fmt.Sprintf("%x", scriptKey)
+		proofMap[key] = a.proofsVec.WithLabelValues(key)
+	}
+
+	// We'll track the size of the proof blob.
+	for key, proof := range proofs {
+		key := fmt.Sprintf("%x", key)
+		a.proofsVec.WithLabelValues(key).
+			Set(float64(len(proof)))
```

Comment:
Looking at this section, I think the file system may be a better way to run this query. We can use `stat` directly, though unless we can somehow batch the system calls, that may also prove to be somewhat slow depending on the underlying virtualization set up.

Reply:
This could be useful: https://pkg.go.dev/io/fs#WalkDir

Reply:
I don't get how looking at the fs directly is feasible? proofs are stored in the db, and in the case of postgres backend that's a service running anywhere, and for sqlite the db could actually be in your file system, but still how can you pin-point which bytes refer to proofs?

Reply:
We also store the proofs as individual files, in addition to the database. See `FileArchiver` in `proof/archive.go`. There we have the script key in the filename (among other things) and the file size would directly correspond to the proof size.

Another alternative would be a DB query that just returns the script key and `length(proof_file)` instead of the full proof, which should already be quite a bit faster.

Reply:
I implemented the query that @guggero recommended, as it seemed more straight forward and was within explored grounds